tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs

```diff
@@ -691,6 +691,145 @@ public async Task GetAuthorizationRequestUrlNoRedirectUriTestAsync()
+
+                Assert.IsNotNull(uri);
+                Dictionary<string, string> qp = CoreHelpers.ParseKeyValueList(uri.Query.Substring(1), '&', true, null);
+                Assert.IsFalse(qp.ContainsKey(Constants.CcsRoutingHintHeader));
+            }
+        }
+
+        private static void AssertCcsHint(Uri uri, string ccsHint)
+        {
+            Assert.IsNotNull(uri);
+            Dictionary<string, string> qp = CoreHelpers.ParseKeyValueList(uri.Query.Substring(1), '&', true, null);
+            Assert.IsTrue(qp.ContainsKey(Constants.CcsRoutingHintHeader));
+            Assert.AreEqual(ccsHint, qp[Constants.CcsRoutingHintHeader]);
+        }
+
+        private static ConfidentialClientApplication CreateCca(MockHttpManager httpManager)
+        {
+            httpManager.AddInstanceDiscoveryMockHandler();
+
+            var app = ConfidentialClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                          .WithAuthority(new Uri(ClientApplicationBase.DefaultAuthority), true)
+                                                          .WithRedirectUri(TestConstants.RedirectUri)
+                                                          .WithClientSecret(TestConstants.ClientSecret)
+                                                          .WithHttpManager(httpManager)
+                                                          .BuildConcrete();
+            return app;
+        }
+
+        [TestMethod]
+        [DataRow(null, TestConstants.TenantId)]
+        [DataRow(TestConstants.OnPremiseUniqueId, null)]
+        [DataRow(null, null)]
+        [DataRow("", "")]
+        public async Task CcsRoutingHint_ClientInfoHintIncorrect_ThrowsAsync(string oid, string tid)
+        {
+            using (var httpManager = new MockHttpManager())
+            {
+                var app = ConfidentialClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                              .WithAuthority(new Uri(ClientApplicationBase.DefaultAuthority), true)
+                                                              .WithRedirectUri(TestConstants.RedirectUri)
+                                                              .WithClientSecret(TestConstants.ClientSecret)
+                                                              .WithHttpManager(httpManager)
+                                                              .BuildConcrete();
+
+                await Assert.ThrowsExceptionAsync<ArgumentException>(() => app.AcquireTokenByAuthorizationCode(TestConstants.s_scope, TestConstants.DefaultAuthorizationCode)
+                    .WithPkceCodeVerifier(string.Empty)
+                    .WithCcsRoutingHint(oid, tid)
+                    .ExecuteAsync()).ConfigureAwait(false);
+            }
+        }
+
+        [TestMethod]
+        [DataRow(null)]
+        [DataRow("")]
+        public async Task CcsRoutingHint_UpnHintIncorrect_ThrowsAsync(string upn)
```

Comment:
The other tests in this file are from https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2713